File: Python/Surrounded Regions.py
class Solution:
    # @param board, a 2D array
    # Capture all regions by modifying the input board in-place.
    # Do not return any value.
	#search from the boundaries, find all neighbour O , these Os can not be surrounded, the search method(BFS or DFS) doesnt matter. 
    def solve(self, board):
		if board==[]:return[]
		colNum = len(board[0])
		lineNum = len(board)
		queue = collections.deque()
		visited = [[False for j in xrange(colNum)] for i in xrange(lineNum)]
		for i in xrange(colNum):
			if board[0][i] == "O":queue.append((0,i))
			if board[lineNum-1][i]=="O":queue.append((lineNum-1,i))
		for i in xrange(lineNum):
			if board[i][0] == "O":queue.append((i,0))
			if board[i][colNum-1] == "O":queue.append((i,colNum-1))
		while queue:
			t = queue.popleft()
			if board[t[0]][t[1]] == "O":board[t[0]][t[1]]="$"
			visited[t[0]][t[1]]=True
			if t[0]+1 < lineNum and board[t[0]+1][t[1]] == "O" and visited[t[0]+1][t[1]]==False:
				queue.append((t[0]+1,t[1]))
			if t[0]-1 > 0 and board[t[0]-1][t[1]] == "O" and visited[t[0]-1][t[1]]==False:
				queue.append((t[0]-1,t[1]))
			if t[1] + 1 < colNum and board[t[0]][t[1] + 1] == 'O' and visited[t[0]][t[1] + 1] == False: 
				queue.append((t[0], t[1] + 1))
			if t[1] - 1 >= 0 and board[t[0]][t[1] - 1] == 'O' and visited[t[0]][t[1] - 1] == False:
				queue.append((t[0], t[1] - 1))
		for i in xrange(lineNum):
			for j in xrange(colNum):
				if board[i][j] == 'O': board[i][j] = 'X'
				if board[i][j] == '$': board[i][j] = 'O'

File: Java/Surrounded Regions.java
class Surrounded Regions{
    public void solve(char[][] board) {
        if (board.length == 0)
            return;
        int rowLen = board.length;
        int colLen = board.length;
        boolean[] visited = new boolean[rowLen][colLen];
        LinkedList<Integer> rowQ = new LinkedList<Integer>();
        LinkedList<Integer> colQ = new LinkedList<Integer>();
        for(int i=0;i<rowLen;i++){
        	if(board[i][0] == 'O'){
        		rowQ.offer(i);
        		colQ.offer(0);
        	}
        	if(board[i][colLen-1] == 'O'){
        		rowQ.offer(i);
        		colQ.offer(colLen-1);
        	}
        }
        for(int i=0;i<colLen;i++){
        	if(board[0][i] == 'O'){
        		rowQ.offer(0);
        		colQ.offer(i);
        	}
        	if(board[rowLen-1][i] == 'O'){
        		rowQ.offer(rowLen-1);
        		colQ.offer(i);
        	}
        }
        while(!rowQ.isEmpty()){
        	int x = rowQ.poll();
        	int y = colQ.poll();
        	if(board[x][y] == 'O')
        		board[x][y] = '$';
        	visited[x][y] = true;
        	if(x+1 < rowLen && visited[x+1][y] == false && board[x+1][y] == 'O'){
        		rowQ.offer(x+1);
        		colQ.offer(y);
        	}
        	if(x-1 > 0 && visited[x-1][y] == false && board[x-1][y] == 'O'){
        		rowQ.offer(x-1);
        		colQ.offer(y);
        	}
        	if(y+1 < colLen && visited[x][y+1] == false && board[x][y+1] == 'O'){
        		rowQ.offer(x);
        		colQ.offer(y+1);
        	}
        	if(y-1 > 0 && visited[x][y-1] == false && board[x][y-1] == 'O'){
        		rowQ.offer(x);
        		colQ.offer(y-1);
        	}
        }
        for(int i=0;i<rowLen;i++){
        	for(int j=0;j<colLen;j++){
        		if(board[i][j] == 'O')
        			board[i][j] = 'X';
        		if(board[i][j] == '$')
        			board[i][j] = 'O';
        	}
        }
    }	
}